miner/multi_collator.go

```diff
@@ -0,0 +1,394 @@
+type collatorBlockState struct {
+	work collatorWork
+	c    *collator
+	done bool
+}
+
+func (c *collatorBlockState) Copy() BlockState {
+	return &collatorBlockState{
+		work: c.work.Copy(),
+		c:    c.c,
+		done: c.done,
+	}
+}
+
+func (bs *collatorBlockState) AddTransactions(sequence types.Transactions, cb AddTransactionsResultFunc) {
+	var (
+		interrupt             = bs.work.interrupt
+		header                = bs.work.env.header
+		gasPool               = bs.work.env.gasPool
+		signer                = bs.work.env.signer
+		chainConfig           = bs.c.chainConfig
+		chain                 = bs.c.chain
+		state                 = bs.work.env.state
+		snap                  = state.Snapshot()
+		curProfit             = big.NewInt(0)
+		coinbaseBalanceBefore = state.GetBalance(bs.work.env.header.Coinbase)
+		tcount                = bs.work.env.tcount
+		err                   error
+		logs                  []*types.Log
+		startTCount           = bs.work.env.tcount
+	)
+	if bs.done {
+		err = ErrAbort
+		cb(err, nil)
+		return
+	}
+
+	for _, tx := range sequence {
+		if interrupt != nil && atomic.LoadInt32(interrupt) != commitInterruptNone {
+			bs.done = true
+			break
+		}
+		if gasPool.Gas() < params.TxGas {
+			log.Trace("Not enough gas for further transactions", "have", gasPool, "want", params.TxGas)
+			err = core.ErrGasLimitReached
+			break
+		}
+		from, _ := types.Sender(signer, tx)
+		// Check whether the tx is replay protected. If we're not in the EIP155 hf
+		// phase, start ignoring the sender until we do.
+		if tx.Protected() && !chainConfig.IsEIP155(header.Number) {
+			log.Trace("encountered replay-protected transaction when chain doesn't support replay protection", "hash", tx.Hash(), "eip155", chainConfig.EIP155Block)
+			err = ErrUnsupportedEIP155Tx
+			break
+		}
+		gasPrice, err := tx.EffectiveGasTip(bs.work.env.header.BaseFee)
+		if err != nil {
+			break
+		}
+		// Start executing the transaction
+		state.Prepare(tx.Hash(), bs.work.env.tcount)
+
+		var txLogs []*types.Log
+		txLogs, err = commitTransaction(chain, chainConfig, bs.work.env, tx, bs.Coinbase())
+		if err == nil {
+			logs = append(logs, txLogs...)
+			gasUsed := new(big.Int).SetUint64(bs.work.env.receipts[len(bs.work.env.receipts)-1].GasUsed)
+			curProfit.Add(curProfit, gasUsed.Mul(gasUsed, gasPrice))
+			tcount++
+		} else {
+			log.Trace("Tx block inclusion failed", "sender", from, "nonce", tx.Nonce(),
+				"type", tx.Type(), "hash", tx.Hash(), "err", err)
+			break
+		}
+	}
+	var txReceipts []*types.Receipt = nil
+	if err == nil {
+		txReceipts = bs.work.env.receipts[startTCount:tcount]
+	}
+	// TODO: deep copy the tx receipts here or add a disclaimer to implementors not to modify them?
+	shouldRevert := cb(err, txReceipts)
+
+	if err != nil || shouldRevert {
+		state.RevertToSnapshot(snap)
+
+		// remove the txs and receipts that were added
+		for i := startTCount; i < tcount; i++ {
+			bs.work.env.txs[i] = nil
+			bs.work.env.receipts[i] = nil
+		}
```

Comment:
This looks strange. No need to `nil` the slice values when you just crop it right below here anyway.